src/java.base/share/classes/java/lang/Process.java

```diff
@@ -149,12 +169,94 @@ public abstract class Process {
+     * This method is idempotent, if this {@code Process} has already been closed
+     * invoking this method has no effect.
+     * <p>
+     * If the data from the process input or error streams is needed, it must be read before
+     * calling this method. The contents of streams that have not been read to end of stream
+     * are lost, they are discarded or ignored.
+     * <p>
+     * If the process exit value of the process is of interest, then the caller must
+     * {@linkplain #waitFor() wait} for the process to terminate before calling this method.
+     * <p>
+     * Streams should be closed when no longer needed.
+     * Closing an already closed stream usually has no effect but is specific to the stream.
+     * If an {@code IOException} occurs when closing a stream it is thrown
+     * when this method returns. Additional {@code IOExceptions}
+     * thrown by closing the remaining streams, if any, are added to the first
+     * {@code IOException} as {@linkplain IOException#addSuppressed suppressed exceptions}.
+     * <p>
+     * After the streams are closed this method {@linkplain #waitFor() waits for} the
+     * process to terminate. If interrupted while {@linkplain #waitFor waiting for termination}
+     * the process is {@linkplain #destroyForcibly() forcibly destroyed}.
+     * The interrupt status will be re-asserted before this method returns and
+     * any {@code IOExceptions} are thrown.
+     * <p>
+     * Try-with-resources example to write text to a process, read back the
+     * response, and close the streams and process:
+     * {@snippet class=ProcessExamples region=example}
+     *
+     * @implSpec
+     * The {@code outputWriter} and {@code outputStream} to the process are closed.
+     * The {@code inputReader} and {@code inputStream} from the process are closed.
+     * The {@code errorReader} and {@code errorStream} from the process are closed.
+     * The process is terminated.
+     * @throws IOException if closing any of the streams throws an exception
+     * @since 26
+     */
+    @Override
+    public void close() throws IOException {
+        synchronized(this) {
+            if (closed) {
+                return;
+            }
+            closed = true;
+            // Close each stream
+            IOException ioe = quietClose(outputWriter != null ? outputWriter : getOutputStream(), null);
+            ioe = quietClose(inputReader != null ? inputReader : getInputStream(), ioe);
+            ioe = quietClose(errorReader != null ? errorReader : getErrorStream(), ioe);
+
+            try {
+                // Wait for process to terminate normally.
+                waitFor();
+            } catch (InterruptedException e) {
+                // Waiting interrupted; forcibly destroy the process
+                destroyForcibly();
+                // Re-assert the interrupt
+                Thread.currentThread().interrupt();
```

Comment:
I think this is still problematic as code after the t-w-r can't be guaranteed that the child has terminated. So I think it needs a loop that ends when waitFor completes without an exception (destroyForcibly would only be called on the first interrupt of course).

Reply:
Thanks for the update. I have sympathy for zombie scenario but it would be too surprising for close to completely normally in that case. So I think the updated proposal looks reasonable and we should see what we can do to improve diagnosability when blocked indefinitely in close.